core/src/main/java/org/lflang/federated/serialization/FedCustomPythonSerialization.java

```diff
@@ -0,0 +1,165 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ***************/
+
+package org.lflang.federated.serialization;
+
+import org.lflang.generator.GeneratorBase;
+import org.lflang.target.Target;
+
+/**
+ * Enables support for custom serialization.
+ *
+ * @author Shulu Li
+ */
+public class FedCustomPythonSerialization implements FedSerialization {
+
+  String customSerializerPackage;
+
+  public FedCustomPythonSerialization(String customSerializerPackage) {
+    this.customSerializerPackage = customSerializerPackage;
+  }
+
+  @Override
+  public boolean isCompatible(GeneratorBase generator) {
+    if (generator.getTarget() != Target.Python) {
+      throw new UnsupportedOperationException("This class only support Python serialization.");
+    }
+    return true;
+  }
+
+  @Override
+  public String serializedBufferLength() {
+    return serializedVarName + ".len";
+  }
+
+  @Override
+  public String serializedBufferVar() {
+    return serializedVarName + ".buf";
+  }
+
+  private String initializeCustomSerializer() {
+    return "if (self->custom_serializer == NULL) {\n"
+        + "PyObject* pName = PyUnicode_DecodeFSDefault(\"%s\");\n"
+            .formatted(customSerializerPackage)
+        + "PyObject* pModule = PyImport_Import(pName);\n"
+        + "Py_DECREF(pName);\n"
+        + "if (PyErr_Occurred()) PyErr_Print();\n"
+        + "if (pModule == NULL) lf_print_error_and_exit(\"Could not load the custom serializer module '%s'.\");\n"
+            .formatted(customSerializerPackage)
+        + "PyObject* SerializerClass = PyObject_GetAttrString(pModule, \"Serializer\");\n"
+        + "if (SerializerClass == NULL) lf_print_error_and_exit(\"Could not find class 'Serializer' in module '%s'.\");\n"
+            .formatted(customSerializerPackage)
+        + "Py_DECREF(pModule);\n"
+        + "self->custom_serializer = PyObject_CallObject(SerializerClass, NULL);\n"
+        + "Py_DECREF(SerializerClass);\n"
+        + "if (self->custom_serializer == NULL) lf_print_error_and_exit(\"Could not instantiate class 'Serializer' in module '%s'.\");\n"
+            .formatted(customSerializerPackage)
+        + "lf_print_log(\"Loaded custom serializer module %s.\\n\");\n"
+            .formatted(customSerializerPackage)
+        + "}\n";
```

Comment:
Perhaps it would be better to provide this as a utility function in one of the .h/.c files, e.g. in the `Python Helper Functions` section of `pythontarget.h/c`?  Also, are you sure about the `Py_DECREF(pModule)` and `Py_DECREF(SerializerClass)`?  Is it possible that garbage collection will free these and make the module unusable?  I am not at all sure about that...

Reply:
Yes that makes sense. I'll move create utility functions in `pythontarget.h/c`. `Py_DECREF()` should be save because `PyImport_Import`, `PyObject_GetAttrString`, `PyObject_CallObject` all increase the reference count so the reference count will not be zero after `Py_DECREF`, hence objects are not garbage collected. However since we don't ever want them to be garbage collected until program termination, it might be better to not do `Py_DECREF`. I'll remove this.

Reply:
Again, I don't see this suggestion addressed in the PR... ???

Reply:
I think this is perhaps because the changes I made are not reflected in this code snippet, because the code snippet here is referencing the code at the time you made the suggestion. There is an "Outdated" tag on the top right of the code snippet indicating this. The 'Files changed' tab on the top shows the total changes of the PR, I believe you can see the changes I made in there.